File: ts-wa/src/pipeline/pipeline.ts
import path from 'path';
import fs from 'fs';
import { keyBy, map, size } from 'lodash';

import { findProjectRoot } from '../utils/project-root';
import { fromPythonFormat, toPythonFormat } from '../utils/adapters';

export type StepStatus = 'SUCCESS' | 'FAILURE' | 'PARTIAL';

export interface PipelineStep {
  name: string;
  function: any;
  input?: any;
  output?: any;
  status?: 'SUCCESS' | 'FAILURE' | 'PARTIAL';
  error?: any;
  key: string;
}

interface PipelineConfig {
  steps: PipelineStep[];
  initialOutputs: Record<string, any>;
  progressDir: string;
  progressFile: string;
}

// Define the necessary interfaces
export interface ProgressIteration {
  iteration: number;
  lastIteration?: number;
  timeStamp: Date;
  steps: ProgressStep[];
}

interface ProgressStep {
  step: number;
  logFile: string;
  status: 'SUCCESS' | 'FAILURE' | 'PARTIAL';
}

interface ProgressData {
  message: string;
  data: any;
  error?: any;
  key: string;
  timeStamp?: Date;
}

function initializeDirectories(progressDir: string, progressFile: string): void {
  fs.mkdirSync(progressDir, { recursive: true });
  if (!fs.existsSync(progressFile)) {
    writePythonicFileData(progressFile, []);
  }
}

function createNewIteration(progressFile: string): ProgressIteration {
  const projectRoot = findProjectRoot();
  const filePath = path.join(projectRoot, progressFile);

  const progressStatus: ProgressIteration[] = readPythonicFileData(progressFile);
  return {
    iteration: progressStatus.length + 1,
    timeStamp: new Date(),
    steps: [],
  };
}

function getLastIteration(progressFile: string): ProgressIteration {
  const progressStatus: ProgressIteration[] = readPythonicFileData(progressFile);
  return progressStatus[progressStatus.length - 1];
}

async function logProgress(
  progressDir: string,
  progressFile: string,
  progressData: ProgressData,
  status: 'SUCCESS' | 'FAILURE' | 'PARTIAL',
  iteration: ProgressIteration
): Promise<void> {
  const progressStatus = readPythonicFileData(progressFile);
  const progressDataLogFile = path.join(progressDir, `${iteration.iteration}.${progressData.key}.log.json`);
  const projectRoot = await findProjectRoot();

  const relativeLogFile = path.relative(projectRoot, progressDataLogFile);

  iteration.steps.push({
    step: iteration.steps.length,
    logFile: relativeLogFile,
    status,
  });

  let existingLogs: ProgressData;
  if (fs.existsSync(progressDataLogFile)) {
    existingLogs = readPythonicFileData(progressDataLogFile);
  }

  existingLogs = { ...progressData, timeStamp: new Date() };
  writePythonicFileData(progressDataLogFile, existingLogs);

  const existingIterationIndex = progressStatus.findIndex(
    (iter: ProgressIteration) => iter.iteration === iteration.iteration
  );
  if (existingIterationIndex !== -1) {
    progressStatus[existingIterationIndex] = iteration;
  } else {
    progressStatus.push(iteration);
  }

  writePythonicFileData(progressFile, progressStatus);
}

function readPythonicFileData(filePath: string): any {
  const fileData = fromPythonFormat(JSON.parse(fs.readFileSync(filePath, 'utf8')));
  return fileData;
}

function writePythonicFileData(filePath: string, data: any): void {
  fs.writeFileSync(filePath, JSON.stringify(toPythonFormat(data), null, 2));
}

// Modified orchestration function to return the last executed step output or previously saved result if all steps were successful
async function orchestrationFunction(
  outputs: Record<string, any>,
  steps: PipelineStep[],
  resumeFromLastSuccessful: boolean
): Promise<any> {
  const previousIteration = getLastIteration(outputs.progressFile);

  // Check if the previous iteration was completely successful
  if (
    resumeFromLastSuccessful &&
    previousIteration &&
    previousIteration.steps.length === steps.length &&
    previousIteration.steps.every((step) => step.status === 'SUCCESS')
  ) {
    console.log('All steps were previously completed successfully. Returning the last step output.');
    const lastStep = previousIteration.steps[previousIteration.steps.length - 1];
    const lastStepOutput = readPythonicFileData(lastStep.logFile);
    return lastStepOutput.data;
  }

  const currentIteration = createNewIteration(outputs.progressFile);

  if (
    previousIteration &&
    (steps.length !== previousIteration.steps.length || previousIteration.steps[steps.length - 1].status !== 'SUCCESS')
  ) {
    for (let s in previousIteration.steps) {
      if (previousIteration.steps[s].status === 'SUCCESS') {
        currentIteration.steps[s] = { ...previousIteration.steps[s] };
        let stepOutput: any = readPythonicFileData(previousIteration.steps[s].logFile);
        outputs = { ...outputs, ...stepOutput.data };
      } else {
        break;
      }
    }
  }

  let lastStepOutput = null;

  for (let i = 0; i < steps.length; i++) {
    const step = steps[i];
    const progressStep = currentIteration.steps.find((s) => s.step === i);

    if (progressStep && progressStep.status === 'SUCCESS') {
      console.log(`Step ${i} (${step.name}) already completed successfully.`);
      outputs = { ...outputs, ...step.output };
      lastStepOutput = step.output;
      continue;
    }

    try {
      console.log(`Executing step ${i} (${step.name})...`);
      const { status, ...stepOutputs } = await step.function(outputs);

      step.status = status;
      step.output = stepOutputs;
      outputs = { ...outputs, ...stepOutputs };
      lastStepOutput = stepOutputs;

      if (step.status !== 'SUCCESS') {
        throw new Error();
      }

      await logProgress(
        outputs.progressDir,
        outputs.progressFile,
        {
          message: `Step ${i} (${step.name}) completed successfully.`,
          data: step.output,
          key: `STEP_${i}_${step.status}_${step.name}`,
        },
        'SUCCESS',
        currentIteration
      );
    } catch (error: any) {
      step.status = step.status || 'FAILURE';
      await logProgress(
        outputs.progressDir,
        outputs.progressFile,
        {
          message: `Step ${i} (${step.name}) failed.`,
          data: { ...step.output },
          error: step.status !== 'PARTIAL' ? { error: [error.message, step.error] } : {},
          key: `STEP_${i}_${step.status}_${step.name}`,
        },
        step.status,
        currentIteration
      );

      console.error(`Step ${i} (${step.name}) failed. Manual intervention required.`);
      throw new Error(`Step ${i} (${step.name}) failed. Manual intervention required.`);
    }

    const progressStatus: ProgressIteration[] = readPythonicFileData(outputs.progressFile);
    const existingIterationIndex = progressStatus.findIndex((iter) => iter.iteration === currentIteration.iteration);
    if (existingIterationIndex !== -1) {
      progressStatus[existingIterationIndex] = currentIteration;
    } else {
      progressStatus.push(currentIteration);
    }
    writePythonicFileData(outputs.progressFile, progressStatus);
  }

  return lastStepOutput;
}

// New function to call the orchestration function with configuration and return the last executed step output
export async function runPipeline(
  steps: PipelineStep[],
  initialOutputs: any,
  progressDir: any,
  progressFile: any,
  resumeFromLastSuccessful: boolean = true
): Promise<any> {
  initializeDirectories(progressDir, progressFile);

  try {
    const lastStepOutput = await orchestrationFunction(
      { ...initialOutputs, progressDir, progressFile },
      steps,
      resumeFromLastSuccessful
    );
    return lastStepOutput;
  } catch (error) {
    console.error('Error in processing: ', error);
    throw error;
  }
}

File: py-wa/app/pipeline/pipeline.py
import os
import json
from pathlib import Path
from typing import List, Any, Dict, Optional

from .types import PipelineStep, PipelineConfig, ProgressIteration, ProgressData
from .progress import get_last_iteration, create_new_iteration, log_progress
from .exceptions import PipelineError, PipelineStepError
from ..utils.logging import setup_logger
from .context import PipelineContext

logger = setup_logger(__name__)

def initialize_directories(progress_dir: str, progress_file: str) -> None:
    """
    Initializes the necessary directories and progress file.

    Creates the progress directory if it doesn't exist (including parent directories).
    If the progress file doesn't exist, it creates it and initializes it with an empty JSON array.

    Args:
        progress_dir: The path to the directory where progress information will be stored.
        progress_file: The path to the file that will store progress information (as a JSON array).

    Returns:
        None
    """
    # Ensure the progress directory exists. Create it recursively if necessary.
    os.makedirs(progress_dir, exist_ok=True)  # Using exist_ok=True is best practice to avoid errors if the directory already exists

    # Check if the progress file exists
    if not os.path.exists(progress_file):
        # If it doesn't exist, create it and initialize with an empty JSON array.
        try:
            with open(progress_file, 'w') as f: # Use 'with open' for automatic file closing (best practice)
                json.dump([], f) # Use json.dump to serialize a Python list to JSON and write to the file
        except IOError as e:
            # Handle potential file writing errors gracefully (best practice)
            print(f"Error writing to progress file '{progress_file}': {e}")
            # Consider raising the exception again or handling it in a way appropriate for your application
            raise

async def orchestrate_pipeline(
    context: PipelineContext,
    outputs: Dict[str, Any],
    steps: List[PipelineStep],
    resume_from_last_successful: bool = True
) -> Any:
    """
    Orchestrate the execution of pipeline steps.
    
    Args:
        context: Pipeline execution context with logger and run_id
        outputs: Initial outputs and configuration
        steps: List of pipeline steps to execute
        resume_from_last_successful: Whether to resume from last successful state
    """
    context.log_pipeline("pipeline_start", total_steps=len(steps))
    
    progress_file = Path(outputs["progress_file"])
    previous_iteration = await get_last_iteration(progress_file)
    
    if previous_iteration:
        context.log_pipeline("previous_iteration_found", step_count=len(previous_iteration.steps))
        
        if (resume_from_last_successful and 
            len(previous_iteration.steps) == len(steps) and 
            all(step.status == "SUCCESS" for step in previous_iteration.steps)):
            
            context.log_pipeline("using_previous_successful_run", iteration=previous_iteration.iteration)
            last_step = previous_iteration.steps[-1]
            last_step_data = json.loads(last_step.log_file.read_text())
            return last_step_data["data"]
    
    current_iteration = await create_new_iteration(progress_file)
    context.log_pipeline("new_iteration_created", iteration=current_iteration.iteration)
    
    last_step_output = None
    
    for i, step in enumerate(steps):
        context.log_step("step_start", i, step.name)
        
        try:
            result = await step.function(outputs, context)
            status = result.pop("status", "SUCCESS")
            step.status = status
            step.output = result
            outputs.update(result)
            last_step_output = result
            
            if step.status != "SUCCESS":
                context.log_step("step_failed", i, step.name, status=status)
                raise PipelineStepError(
                    message="Step execution failed",
                    step_name=step.name,
                    step_number=i,
                    step_context={"status": status}
                )
            
            progress_data = ProgressData(
                message=f"Step {i} ({step.name}) completed successfully.",
                data=step.output,
                key=f"STEP_{i}_{step.status}_{step.name}"
            )
            
            context.log_step("complete", i, step.name, status="SUCCESS")
            await log_progress(
                Path(outputs["progress_dir"]),
                progress_file,
                progress_data,
                "SUCCESS",
                current_iteration
            )
            
        except Exception as e:
            step.status = step.status or "FAILURE"
            context.log_step("failed", i, step.name, status=step.status, error=str(e))
            
            error_data = ProgressData(
                message=f"Step {i} ({step.name}) failed.",
                data=step.output or {},
                error={"error": [str(e), step.error]} if step.status != "PARTIAL" else {},
                key=f"STEP_{i}_{step.status}_{step.name}"
            )
            
            await log_progress(
                Path(outputs["progress_dir"]),
                progress_file,
                error_data,
                step.status,
                current_iteration
            )
            
            raise PipelineError(
                message=f"Step {i} ({step.name}) failed. Manual intervention required.",
                context={"step": i, "name": step.name}
            )
    
    context.log_pipeline("pipeline_complete")
    return last_step_output

async def run_pipeline(
    context: PipelineContext,
    steps: List[PipelineStep],
    initial_outputs: Dict[str, Any],
    progress_dir: Path,
    progress_file: Path,
    resume_from_last_successful: bool = True
) -> Any:
    """
    Run the pipeline with the given configuration.
    
    Args:
        steps: List of pipeline steps to execute
        initial_outputs: Initial data for the pipeline
        progress_dir: Directory for progress tracking
        progress_file: File for storing progress state
        resume_from_last_successful: Whether to resume from last successful state
        
    Returns:
        Any: Output from the last successful step
        
    Raises:
        PipelineError: If pipeline execution fails
    """
    context.log_pipeline("pipeline_init")
    
    try:
        progress_dir.mkdir(parents=True, exist_ok=True)
        if not progress_file.exists():
            context.log_pipeline("create_progress_file")
            progress_file.write_text(json.dumps([]))
    except Exception as e:
        context.log_pipeline("init_failed", error=str(e))
        raise PipelineError("Directory initialization failed", {"error": str(e)})
    
    try:
        outputs = {
            **initial_outputs,
            "progress_dir": str(progress_dir),
            "progress_file": str(progress_file)
        }
        return await orchestrate_pipeline(context, outputs, steps, resume_from_last_successful)
    except Exception as e:
        context.log_pipeline("failed", error=str(e))
        raise

